### PR TITLE
#497 Agenda.removeAll() on destroy

### PR DESCRIPTION
**Pull request by amihaiemil**
PR for #497

Added and tested method ``Agenda.removeAll(...)``. The alternative was to call ``Agenda.item()`` and run the xpath query inside ``destroy.groovy``, but I think it's better to keep XML classes out of the groovy scripts, let them be encapsulated.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#806](https://github.com/zerocracy/farm/issues/806) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#806](https://github.com/zerocracy/farm/issues/806) is assigned to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), here is [why](https://www.0crat.com/footprint/C3NDPUA8L/721821355). The budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4). Please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

File: src/test/java/com/zerocracy/pmo/AgendaTest.java [hunk 1]
```diff
@@ -41,4 +42,43 @@ public void addsAndRemovesAgenda() throws Exception {
...
     }
 
+    /**
+     * Agenda can remove all the orders.
+     * @throws Exception If something goes wrong.
+     */
+    @Test
+    public void removesAllOrders() throws Exception {
+        final Agenda agenda = new Agenda(new FkProject(), "mihai").bootstrap();
+        agenda.add("gh:test2/test#1", "REV");
```
Comment (carlosmiranda) [inline src/test/java/com/zerocracy/pmo/AgendaTest.java]: @amihaiemil Qulice PMD check is complaining about `"REV"` duplicate string.

File: src/test/java/com/zerocracy/pmo/AgendaTest.java [hunk 2]
```diff
@@ -41,4 +42,43 @@ public void addsAndRemovesAgenda() throws Exception {
...
+        agenda.add("gh:test2/test#2", "QA");
+        agenda.add("gh:test2/test#3", "DEV");
+        MatcherAssert.assertThat(agenda.jobs(), Matchers.not(0));
+        agenda.removeAll();
+        MatcherAssert.assertThat(agenda.jobs(), Matchers.hasSize(0));
```
Comment (carlosmiranda) [inline src/test/java/com/zerocracy/pmo/AgendaTest.java]: @amihaiemil let's use `Matchers.empty()` here.

File: src/test/java/com/zerocracy/pmo/AgendaTest.java [hunk 3]
```diff
@@ -41,4 +42,43 @@ public void addsAndRemovesAgenda() throws Exception {
...
+     * Agenda can remove the only order.
+     * @throws Exception If something goes wrong.
+     */
+    @Test
+    public void removesSoleOrder() throws Exception {
+        final Agenda agenda = new Agenda(new FkProject(), "john").bootstrap();
+        agenda.add("gh:test3/test#1", "ARC");
+        MatcherAssert.assertThat(agenda.jobs(), Matchers.hasSize(1));
+        agenda.removeAll();
+        MatcherAssert.assertThat(agenda.jobs(), Matchers.hasSize(0));
```
Comment (carlosmiranda) [inline src/test/java/com/zerocracy/pmo/AgendaTest.java]: @amihaiemil same here, let's use `Matchers.isEmpty()`

File: src/test/java/com/zerocracy/pmo/AgendaTest.java [hunk 4]
```diff
@@ -41,4 +42,43 @@ public void addsAndRemovesAgenda() throws Exception {
...
+    }
+
+    /**
+     * Agenda can "remove" orders if it's empty.
+     * @throws Exception If something goes wrong.
+     */
+    @Test
+    public void removesOrdersFromEmptyAgenda() throws Exception {
+        final Agenda agenda = new Agenda(new FkProject(), "jane").bootstrap();
+        MatcherAssert.assertThat(agenda.jobs(), Matchers.hasSize(0));
```
Comment (carlosmiranda) [inline src/test/java/com/zerocracy/pmo/AgendaTest.java]: @amihaiemil same here, let's use `Matchers.isEmpty()`

File: src/test/java/com/zerocracy/pmo/AgendaTest.java [hunk 5]
```diff
@@ -41,4 +42,43 @@ public void addsAndRemovesAgenda() throws Exception {
...
+        agenda.removeAll();
+        MatcherAssert.assertThat(agenda.jobs(), Matchers.hasSize(0));
```
Comment (carlosmiranda) [inline src/test/java/com/zerocracy/pmo/AgendaTest.java]: @amihaiemil same here, let's use `Matchers.isEmpty()`

File: src/test/java/com/zerocracy/pmo/AgendaTest.java [hunk 6]
```diff
@@ -41,4 +42,43 @@ public void addsAndRemovesAgenda() throws Exception {
...
+     * Agenda can remove all the orders.
+     * @throws Exception If something goes wrong.
+     */
+    @Test
+    public void removesAllOrders() throws Exception {
+        final Agenda agenda = new Agenda(new FkProject(), "mihai").bootstrap();
+        agenda.add("gh:test2/test#1", "REV");
+        agenda.add("gh:test2/test#2", "QA");
+        agenda.add("gh:test2/test#3", "DEV");
+        MatcherAssert.assertThat(agenda.jobs(), Matchers.not(0));
```
Comment (carlosmiranda) [inline src/test/java/com/zerocracy/pmo/AgendaTest.java]: @amihaiemil This assertion will always pass, won't it? `agenda.jobs()` will never be equal to the value `0`. Why not use `Matchers.hasSize(3)`?

**Comment by amihaiemil** [comment]
@carlosmiranda fixed. Had to add that PMD escaping, since the roles are limited and I could not avoid repetition. Also added escaping for ``3 is a magic number``, I saw it's done throughout the project.

**Comment by carlosmiranda** [comment]
@yegor256 this is good to merge

**Comment by yegor256** [comment]
@rultor merge

**Comment by amihaiemil** [comment]
@yegor256 can we try to merge again? That test doesn't seem to have anything to do with these changes.

**Comment by amihaiemil** [comment]
@yegor256 please try to merge again

**Comment by amihaiemil** [comment]
@yegor256 ping

**Comment by yegor256** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#806](https://github.com/zerocracy/farm/issues/806) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256)

**Comment by ypshenychka** [comment]
@carlosmiranda According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

Only two issues were found during code review. 
Please confirm that you'll try to find at least three problems while future reviews.

**Comment by ypshenychka** [comment]
@carlosmiranda Thanks!

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 points just awarded to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda)

**Comment by 0crat** [comment]
Quality review completed: +8 points just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**